<a href="https://colab.research.google.com/github/negatively/CaseStudy-DataCamp-python/blob/main/credit_card_fraud.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Pendahuluan
Dalam projek ini kita akan menggunakan beberapa model prediksi untuk melihat bagaimana akurasi model dalam mendeteksi transaksi itu termasuk normal atau fraud. Berdasarkan deskripsi dari dataset, feature telah dinamai dengan V untuk privasi dan telah discaling. Namun, kita masih dapat menganalisa aspek-aspek penting dari dataset.

## Tujuan
* Memahami distribusi dataset dari sebagian data yang telah diberikan kepada kita
* Membuat sub-dataframe dengan rasio 50-50 untuk Fraud dan non-Fraud.
* Menentukan classifier yang akan digunakan dan menentukan mana yang memiliki akurasi tertinggi
* Membuat Neural Network dan membandingkan hasil akurasi dengan classifier terbaik
* Memahami kesalahan yang biasa terjadi dengan imbalanced dataset


## Data Understanding
Hal pertama yang perlu kita lakukan adalah mengumpulkan informasi terkait dataset kita. Perlu diingat, kecuali kolom transaction dan amount kita tidak tahu isi dari kolom lain. Yang kita tahu kolom-kolom tersebut telah di scaling.


In [2]:
# import libraries
import numpy as np
import pandas as pd
import tensorflow as tf
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.manifold import TSNE
from sklearn.decomposition import PCA, TruncatedSVD
import matplotlib.patches as mpatches
import time
from zipfile import ZipFile

#Libray Classifier
from sklearn.model_selection import train_test_split, KFold, StratifiedKFold
from sklearn.pipeline import make_pipeline
from imblearn.pipeline import make_pipeline as imbalanced_make_pipeline
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import NearMiss
from imblearn.metrics import classification_report_imbalanced
from sklearn.metrics import precision_score, recall_score, f1_score, roc_auc_score, accuracy_score, classification_report
from collections import Counter
import warnings
warnings.filterwarnings("ignore")

/usr/local/lib/python3.7/dist-packages/sklearn/externals/six.py:31: FutureWarning: The module is deprecated in version 0.21 and will be removed in version 0.23 since we've dropped support for Python 2.7. Please rely on the official version of six (https://pypi.org/project/six/).
  "(https://pypi.org/project/six/).", FutureWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.neighbors.base module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.neighbors. Anything that cannot be imported from sklearn.neighbors is now part of the private API.
  warnings.warn(message, FutureWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.metrics.classification module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be import

In [3]:
# Set Kaggle API Permission
! pip install -q kaggle
from google.colab import files
files.upload()
! mkdir ~/.kaggle
! cp kaggle.json ~/.kaggle/
! chmod 600 ~/.kaggle/kaggle.json

Saving kaggle.json to kaggle.json


In [4]:
# Get Dataset From Kaggle
!kaggle datasets download -d mlg-ulb/creditcardfraud

 92% 61.0M/66.0M [00:00<00:00, 61.1MB/s]
100% 66.0M/66.0M [00:00<00:00, 122MB/s] 


In [5]:
path = 'creditcardfraud.zip'
with ZipFile(path, 'r') as zip_ref:
    zip_ref.extractall('working')

In [6]:
# Load dataset
credit = pd.read_csv('working/creditcard.csv')

In [7]:
credit.head(10)

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0
5   2.0 -0.425966  0.960523  1.141109  ...  0.253844  0.081080    3.67      0
6   4.0  1.229658  0.141004  0.045371  ...  0.034507  0.005168    4.99      0
7   7.0 -0.644269  1.417964  1.074380  ... -1.206921 -1.085339   40.80      0
8   7.0 -0.894286  0.286157 -0.113192  ...  0.011747  0.142404   93.20      0
9   9.0 -0.338262  1.119593  1.044367  ...  0.246219  0.083076    3.68      0

[10 rows x 31 columns]

In [8]:
credit.describe()

Time            V1  ...         Amount          Class
count  284807.000000  2.848070e+05  ...  284807.000000  284807.000000
mean    94813.859575  3.919560e-15  ...      88.349619       0.001727
std     47488.145955  1.958696e+00  ...     250.120109       0.041527
min         0.000000 -5.640751e+01  ...       0.000000       0.000000
25%     54201.500000 -9.203734e-01  ...       5.600000       0.000000
50%     84692.000000  1.810880e-02  ...      22.000000       0.000000
75%    139320.500000  1.315642e+00  ...      77.165000       0.000000
max    172792.000000  2.454930e+00  ...   25691.160000       1.000000

[8 rows x 31 columns]

In [14]:
credit.isna().sum().max()

0

In [27]:
print("Dalam dataset terdapat data Non-Fraud sebanyak", round(credit["Class"].value_counts()[0]/len(credit.Class)*100, 2), "%")
print("Dalam dataset terdapat data Fraud sebanyak", round(credit.Class.value_counts()[1]/len(credit)*100, 2), "%")

Dalam dataset terdapat data Non-Fraud sebanyak 99.83 %
Dalam dataset terdapat data Fraud sebanyak 0.17 %


**Rangkuman :**
* Besar transaksi yang ada relatif kecil dengan rata-rata 88 USD
* Tidak terdapat null values, sehingga kita tidak perlu untuk mengatasinya
* Kebanyakan transaksi adalah Non-fraud(99.83%) sedangkan Fraud(0.17%)